Fixed #1706

mma5997 opened the pull request:
This makes the request body inline as we're already doing it for responses.
This will help flatten api to leverage the inline schema for this resolved request body which has an inline schema inside it.

mma5997 commented:
Just saw that 2 tests failed. 
Hmm seems we'd need to fix those 2.
But looks everything else is smooth.

mma5997 commented:
updated the test files as per the primary changes done in this PR.

mma5997 commented:
Hi @gracekarina 
I am done with all the changes, please can you review and help proceeding with the workflows ahead.
Thanks for your time :)

gracekarina commented:
@mma5997 Hi!, sorry it took so long for me to reply, I was reviewing the changes and they seem ok, but we should not change a functionality that could break so many dependent users of parser's current behavior, so, what we can do is add an option in ParseOptions like flatten, resolveFully, etc,  and if anyone needs it, configure parser with the option set to true, and that's it. https://github.com/swagger-api/swagger-parser/blob/master/modules/swagger-parser-core/src/main/java/io/swagger/v3/parser/core/models/ParseOptions.java#L6 if you need any help adding the option let me know. Thanks for this PR!!!

mma5997 commented:
No issues, I was eagerly waiting for your response :) 
Thanks for the suggestion @gracekarina  , you're right, change in behaviour would impact the backward compatibility and the existing user-base. 

So, I've tried adding a new parseOption to toggle the resolution of the requestBody accordingly.
Please can you help reviewing it. Please do let me know your thoughts.

PS:
Maybe later I'll need to add separate unit test for this scenario. 
Also I had made a similar parseOption for response resolution (which happens by default so its default value is true) which might be on similar lines to requestBody. But seems there's a test - `io.swagger.v3.parser.processors.OperationProcessorTest.testProcessOperation(List<Parameter>, List<Parameter>, ApiResponse, ApiResponse)` failing due to some reasons, and I thus made the changes only for requestBody apparently.

gracekarina commented:
Hi, @mma5997  thanks for this approach, at the moment I have to do a thorough review of this solution, as it seems, it's still putting inline the requestBodies, and the option is not being set in the test. I will work on a different branch and commit the changes, they will be in the next release.

mma5997 commented:
Thanks @gracekarina 😃 , sure!
just to clarify and summarize once about the final changes [here](https://github.com/swagger-api/swagger-parser/pull/1707/files)

1. I have just added a member var `resolveRequestBody` in ParseOptions class, whose default value is `false` because its the normal default behaviour that, the parser doesn't put the request body inline. So the existing parser user-base will not face any problem. The setter for that var (`resolveRequestBody`) will help setting the toggle for putting the requestBody inline which solves #1706.
2. In OperationProcessor, I've added the check to see if the `resolveRequestBody` is set to `true` using the getter for the same and then put requestBody inline IFF its set to true.
3. Added the getter in ResolverCache class for getting the ParseOptions instance which will help us check if `resolveRequestBody` is set to `true` or not in the above point 2.

The `resolveRequestBody` ParseOption is not needed to be set explicitly in the test or anywhere else because the default behaviour is kept intact by defaulting  `resolveRequestBody` to `false` as aforementioned.
I haven't added any unit test for `resolveRequestBody` set to true, yet.

Thanks again 😄 !

gracekarina commented:
Hi can you please add the test for the option set to true? thanks

mma5997 commented:
Hello @gracekarina,
Thanks for your time, I've added a simple test, please review :)

frantuma commented on modules/swagger-parser-core/src/main/java/io/swagger/v3/parser/core/models/ParseOptions.java:
```diff
@@ -36,6 +37,18 @@ public boolean isResolveFully() {
+     */
+	public void setResolveRequestBody(boolean resolveRequestBody) {
+		this.resolveRequestBody = resolveRequestBody;
+	}
```
Would need to use spaces instead of tabs and correct indentation

mma5997 replied:
Done!

frantuma commented on modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/ResolverCache.java:
```diff
@@ -68,11 +68,10 @@ public class ResolverCache {
 
 
-    /*
+	/*
```
Would need to use spaces instead of tabs and correct indentation

mma5997 replied:
Done!

frantuma commented on modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/ResolverCache.java:
```diff
@@ -394,4 +393,8 @@ public Map<String, String> getExternalFileCache() {
         return Collections.unmodifiableMap(renameCache);
     }
+    
+    public ParseOptions getParseOptions() {
+		return parseOptions;
+	}
```
Would need to use spaces instead of tabs and correct indentation

mma5997 replied:
Done!

frantuma commented on modules/swagger-parser-v3/src/main/java/io/swagger/v3/parser/processors/OperationProcessor.java:
```diff
@@ -40,8 +39,19 @@ public void processOperation(Operation operation) {
+                	requestBody = resolvedRequestBody;
+                	operation.setRequestBody(resolvedRequestBody);
+                }
+        	}
```
Would need to use spaces instead of tabs and correct indentation

mma5997 replied:
Done!

frantuma commented on modules/swagger-parser-v3/src/test/resources/issue-1706/SimpleRequestResponseRef.json:
```diff
@@ -0,0 +1,87 @@
+{
```
Would need to use spaces instead of tabs

mma5997 replied:
done

mma5997 replied:
Done!

frantuma replied:
@mma5997 it seems that json file still has tabs instead of spaces. Sorry about that, it's to get PRs as clean as we can

mma5997 replied:
Thanks for pointing it out @frantuma. 
Sorry, my bad, Eclipse formatting likes tabs it seems XD. No issues at all, I agree with you, it should be clean definitely. I have fixed it.

frantuma reviewed (commented):
Hi @mma5997 thanks a lot! can you possibly just fix the indent/tabs things mentioned in review comments?

mma5997 commented:
Thanks @frantuma, 
please can you review once and let me know if anything's needed, I'll be more than happy :)

frantuma commented:
Thanks for your contribution, merged!
